README.md
<p align="center"><a href="https://tridiamond.tech" target="_blank" rel="noopener noreferrer"><img width="100" src="https://img-blog.csdnimg.cn/20200930013332450.png" alt="TriDiamond logo"></a></p>

<h1 align="center">Hexo Theme: Obsidian</h1>

<div align="center">

v1.x | ᴅᴇsɪɢɴᴇᴅ & ᴄᴏᴅᴇᴅ ʙʏ ᴛʀɪᴅɪᴀᴍᴏɴᴅ <br>
A dark Hexo theme, it's responsive, simple but elegant.

  <p align="center">
    <img src="https://img.shields.io/github/v/release/TriDiamond/hexo-theme-obsidian">
    <img src="https://img.shields.io/github/release-date/TriDiamond/hexo-theme-obsidian">
    <img src="https://img.shields.io/github/license/TriDiamond/hexo-theme-obsidian">
    <a href="https://gitter.im/TriDiamond/hexo-theme-obsidian?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge"><img src="https://badges.gitter.im/TriDiamond/hexo-theme-obsidian.svg"></a>
  </p>

**[PREVIEW](https://obsidian.tridiamond.tech/)** | **[CHANGES](https://github.com/TriDiamond/hexo-theme-obsidian/blob/master/CHANGELOG.md)**

🇨🇳 **[中文文档](https://github.com/TriDiamond/hexo-theme-obsidian/blob/master/README_CN.md)** |
**[更变日志](https://github.com/TriDiamond/hexo-theme-obsidian/blob/master/CHANGELOG_CN.md)**

</div>

> I am working on a brand new theme called [`Aurora`](https://github.com/auroral-ui/hexo-theme-aurora), therefore this theme currently is not actively maintained. If you are from my blog website, the theme displayed there is the new theme.

<details>
  <summary>View new theme screenshots</summary>

![](https://img-blog.csdnimg.cn/202103280030531.png)

![](https://img-blog.csdnimg.cn/20210328003140590.png)

</details>

![screenshot](https://res.cloudinary.com/tridiamond/image/upload/v1573323147/blog/A-Obsidian-full_ubmo0d.png)

<details>
<summary>More screen shots</summary>

## Categories Page

![screenshot](https://res.cloudinary.com/tridiamond/image/upload/v1573148012/blog/A-Obsidian-categories_mgdti7.png)

## Categories Leveled

![screenshot](https://res.cloudinary.com/tridiamond/image/upload/v1573148016/blog/A-Obsidian-categories-level_xtxty9.png)

## Archives

![screenshot](https://res.cloudinary.com/tridiamond/image/upload/v1573323148/blog/A-Obsidian-archives_ffpwf9.png)

## Articles

![screenshot](https://res.cloudinary.com/tridiamond/image/upload/v1573148016/blog/A-Obsidian-articles_wlsu2v.png)

</details>

## Install

```bash
git clone https://github.com/TriDiamond/hexo-theme-obsidian.git obsidian
```

## Usage

### Activate Theme

Open `Hexo` config file `_config.yml`, set theme to `obsidian`

```yaml
---
theme: obsidian
```

### Installing required packages

First open your terminal, and `cd` in the theme folder `obsidian`.

```shell
cd themes/obsidian
```

Then run `npm` to install all required packages for the theme.

```shell
npm install
```

### Theme settings

Open `/themes/obsidian/_config.yml` can change the theme configs

<details>
<summary>Full detail config</summary>

```yaml
#! ---------------------------------------------------------------
#!  ▄▀▄ █▀▄ ▄▀▀ ▀ █▀▄ ▀ ▄▀▄ █▄░█
#!  █░█ █▀█ ░▀▄ █ █░█ █ █▀█ █░▀█
#!  ░▀░ ▀▀░ ▀▀░ ▀ ▀▀░ ▀ ▀░▀ ▀░░▀
#! ᴅᴇsɪɢɴᴇᴅ & ᴄᴏᴅᴇᴅ ʙʏ ᴛʀɪᴅɪᴀᴍᴏɴᴅ
#! ---------------------------------------------------------------
#! Theme Obsidian - A dark Hexo theme, it's responsive, simple but elegant.
#! Thanks to and inspirated by CSS-Tricks
#! @author: TriDiamond
#! @github: https://github.com/TriDiamond
#! @changelogs: https://github.com/TriDiamond/hexo-theme-obsidian/blob/master/CHANGELOG.md
#! @feedbacks: https://github.com/TriDiamond/hexo-theme-obsidian/issues/new
#! ---------------------------------------------------------------

#! ---------------------------------------------------------------
#!  █▄░█ ▄▀▄ ▀█▀ █▀▀ ▄▀▀
#!  █░▀█ █░█ ░█░ █▀▀ ░▀▄
#!  ▀░░▀ ░▀░ ░▀░ ▀▀▀ ▀▀░ v1.x
#! ---------------------------------------------------------------
#! Since version `v1.3.5` code block highlight has changed from `highlight.js` to
#! `codemirror`, but the Hexo default highlight still need to be disabled to take affect.
#! !! So Please make sure you disabled Hexo's default highlight setting !!
#! ---------------------------------------------------------------
#! Rebuild files after you changed the config:
#! `hexo clean && hexo g`
#! ---------------------------------------------------------------
#! highlight:
#!  enable: false
#!  line_number: true
#!  auto_detect: true
#!  tab_replace:
#! ---------------------------------------------------------------

# Theme version
# !! DO NOT EDIT THIS !!
version: '1.4.9.3'

# ---------------------------------------------------------------
# Theme default settings
# ---------------------------------------------------------------

# Menu setting | format = title_name: link_url
menu:
  PAGE: /page

# Page custom titles
page_titles:
  categories: 'Categories'
  tags: 'Tags'
  archives: 'Archived'

# Social media links
socials:
  github: ""
  twitter: ""
  stackoverflow: ""

# Enable table of content
TOC: true

# Home page first post default cover image, default use cover
welcome_cover: /img/cover.jpg

# Article default cover image
cover: /img/welcome-cover.jpg

# Avatar
avatar: https://s2.ax1x.com/2019/09/19/nLtSiD.png

# Disable default scrollbar
scrollbar: true

# Website keywords used for SEO
keywords: TriDiamond Obsidian

# Website slogans
descriptionOne: 'Think like an artist, develop like an artisan'
descriptionTwo: '艺术家思维去思考问题，工匠创造精神去开发'

# If you use google analytics, please fill in the ID
google_analytics:

# Website favicon
favicon: /img/favicon.png

# rss file
rss: atom.xml

# ---------------------------------------------------------------
# Article Music settings
# ---------------------------------------------------------------

# Auto play article audio
autoplay: false

# default mp3 file
mp3:
  - statics/chengdu.mp3

# ---------------------------------------------------------------
# Plugins
# ---------------------------------------------------------------

# Gitalk comment plugin
# see https://github.com/gitalk/gitalk
gitalk:
  enable: true
  autoExpand: false
  clientID: ''
  clientSecret: ''
  repo: ''
  owner: ''
  admin: ['']
  # Ensure uniqueness and length less than 50
  id: location.pathname
  # Facebook-like distraction
  proxy: https://cors-anywhere.azm.workers.dev/https://github.com/login/oauth/access_token
  # Custom proxy server

# Valine comment plugin (recommended!)
# see https://valine.js.org/quickstart.html
valine:
  enable: false
  app_id:
  app_key:
  notify: false
  verify: false
  avatar: 'mp'
  placeholder: 'Leave your throughs behind~'
  visitor: true

# Enable symbols-count-time plugin
# see https://github.com/theme-next/hexo-symbols-count-time
symbols_count_time:
  enable: true
  wordCount: true
  readCount: true
  awl: 4
  wpm: 275
  suffix: mins.

# Enable html truncate
# see https://github.com/TriDiamond/hexo-html-truncate
html_truncate:
  enable: true
  # Characters kept for posts
  postLength: 250
  # Characters kept for cover posts
  coverLength: 100
  ellipsis: '...'
  # Excluding html tags
  excludes: ['img']
  # Characters count including white spaces
  keepWhitespaces: true
  # Reserving the last complete word, without breaking the word
  reserveLastWord: true

# Enable Busuanzi statistic plugin
# see http://ibruce.info/2015/04/04/busuanzi/
busuanzi:
  enable: true

# Enable social media sharing
# see https://github.com/overtrue/share.js/
sharejs:
  enable: true
  disabled: 'facebook,douban,linkedin,diandian,tencent,google'

# Use codemirror instead of highlight js for better colors
# if you want more language support check the modes list at official website
# modes see: https://codemirror.net/mode/
# see https://codemirror.net/
codemirror:
  modes: ['javascript', 'css', 'xml', 'htmlmixed', 'clike', 'php', 'shell', 'python']

# Enabling mathjax support
mathjax: true

# This function is only used for those that server is located in China. If your server is located in outside China, you do not need to add the following code.
#beianloc:
beian: 
# Enable
  enable: true
# Beian location，for example：粤ICP备
  beianloc:
# Beian ID，for example：2021022134号。
  beianid:

# Police Beian
police:
# Enable
  enable: true
# Beian location，for example: 粤公网安备
  beianloc:
# Beian ID, for example: 44010602009049号
  beianid:
# Beian Icon, for example: /img/beian.png
  beianicon:

# Custom mouse style
mouse:
# Enable custom mouse style
  enable: true
# Default style
  default: /default.png
# Hover sytle
  pointer: /hover.png

# APlayer, now only support netease music list
aplayer:
  # Enable aplayer
  enable: true
  # Music list ID
  musiclist: ""
  
# Live2D
l2d:
  # Enable live2d
  enable: true
```

</details>

### Comments

You can choose from using Valine or Gitalk

- If you are using `Valine`, you need to turn off `busuanzi`,because Valine has article view count support.
- If you are using `Gitalk`, you may turn on `busuanzi`, to have article view count enable.

### Code block style

> Since version `v1.3.5` code block highlight has changed from `highlight.js` to
> `codemirror`, Hexo default highlight need to be disabled to take affect.
> !! So Please make sure you disabled Hexo's default highlight setting !!

Hexo configuration file `_config.yml` located at the ROOT path of your blog.

> ⚠️ Please note do not make these changes in the theme's config file,
> changing configs in the theme config file `_config.yml`
> will not work properly.

```yaml
---
highlight:
  enable: false
  line_number: true
  auto_detect: true
  tab_replace:
```

Codemirror settings

```yaml
# Use codemirror instead of highlight js for better color syntax
# if you want more language support check the modes list at official website
# modes see: https://codemirror.net/mode/
# see https://codemirror.net/
codemirror:
  modes: ['javascript', 'css', 'xml', 'htmlmixed', 'clike', 'php', 'shell', 'python']
```

Rebuild files after you changed the config

```bash
hexo clean && hexo g
```

### Preview character length

This post parameter changes how many characters are kept for the preview at home page.

```markdown
title: My awesome title
date: 2019-07-14 18:38:45
categories: - Category1 - Category2
tags: - Tag1 - Tag2
mp3: http://domain.com/awesome.mp3
cover: http://domain.com/awesome.jpg
preview: 300
```

### Post template

Post template settings

```markdown
title: My awesome title
date: 2019-07-14 18:38:45
categories: - Category1 - Category2
tags: - Tag1 - Tag2
mp3: http://domain.com/awesome.mp3
cover: http://domain.com/awesome.jpg

author: 3rdparty author
socials: 
  page: https://example
avatar: /example.jpg
slogan: example

socials:
    page: 
        link: social platform link
        name: the iconfont you want to use, for example: icon-bilibili-fill
        path: the css file of your iconfont, for example: /css/font_bilibili/iconfont.css
    page2: 
        link: 
        name: 
        path: 
    page3: 
        link: 
        name: 
        path: 
```

### Create categories page

Run Hexo command to generate categories page

```bash
hexo new page categories
```

Categories template

```yaml
title: categories
date: 2019-07-14 12:39:04
type: 'categories'
```

> Theme will auto generate categories contents.

### Create tags page

Run Hexo command to generate tags page

```bash
hexo new page tags
```

Tags page template

```
title: tags
date: 2014-12-22 12:39:04
type: "tags"
```

> Theme will auto generate tags contents.

### MathJax Support

Default enables mathjax support, can change it to `false` to turn it off.

```yaml
# Enabling mathjax support
mathjax: true
```

## Update Theme

> Please backup your `_config.yml` file before update

```bash
cd themes/osidian
git pull
```

## Using the search engine

### Installation guide

Install by npm

```bash
$ npm install hexo-generator-search --save
```

Adding the plugin config in your root `_config.yml`

```yaml
search:
  path: search.xml
  field: post
  content: true
```

Regenerate files

```bash
hexo clean && hexo g
```

## Related

I have made a [`Typro`](https://typora.io/) theme [`Obsidian`](https://github.com/TriDiamond/typro-theme-obsidian) especially for this blog theme.

## Feedback

> Please check the [ISSUES](https://github.com/TriDiamond/hexo-theme-obsidian/issues) before try to use the following contact methods!
> Maybe the question you have had being asked by others or it's already being answered. Thanks!

You can give me feedback or issue you have through the following methods:

- [Create a issue!](https://github.com/TriDiamond/hexo-theme-obsidian/issues/new)
- Joint QQGroup `909955326`
- [Join Telegram Group](https://t.me/joinchat/R2m4eho2lbcHLR7nDvxd6A)

## Donation

Are you **enjoying this project** ? 👋

You can express your ❤️ by _buying me a coffee_ ☕️ to keep this project **maintained and stay alive**, I would ❤️ to **dedicate more time and effort** on it!

If there are enough coffee ☕️ I would like to become a **full time open source developer**! Keep producing more _awesome themes and applications to the world_!

However you could just **sharing this project with your friends**, that would _help me a lot as well_! 👊

Thanks for your love in advance! ☀️

|                                                                                                                Donating using Paypal                                                                                                                 |
| :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <a href="https://www.buymeacoffee.com/tridiamond" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/v2/default-yellow.png" alt="Buy Me A Coffee" height="60" width="217" style="height: 60px !important;width: 217px !important;" ></a> |

> All donator will be enlisted as Coffee Supporter and displayed in the Sponsor list in this section.

---

|                                         Wechat 微信支付                                         |                                        Alipay 支付宝支付                                        |
| :---------------------------------------------------------------------------------------------: | :---------------------------------------------------------------------------------------------: |
| <img src="https://img-blog.csdnimg.cn/20210330175112304.png" alt="Buy Me A Coffee" width="150"> | <img src="https://img-blog.csdnimg.cn/20210330175153827.png" alt="Buy Me A Coffee" width="150"> |

> 所有赞助人将被列为咖啡支持者，并在本节的赞助人名单中显示。

## Join the Community

<a href="https://discord.gg/VC7CrYfds5" target="_blank"><img src="https://discordapp.com/api/guilds/801943105913225246/widget.png?style=banner3" alt="ObsidiaNext Community Discord Server"></a>


CHANGELOG.md
# Changelog

[中文更变日志](https://github.com/TriDiamond/hexo-theme-obsidian/blob/master/CHANGELOG_CN.md)

## 🚀 v1.4.9.3 (Jul 20, 2022)

- Add custom abstract for each post.
    - You can still use ``preview`` to show your posts' headings as abstract.
    - It will show your ``preview`` in default.
- Now you can add the following in your front-meta to show your custom abstract:

```yaml
---
...
# This will show in your post page, keep it short.
abstract: ""
# This is your detail abstract.
abstracts: ""
...
```

## 🚀 v1.4.9.3 (Jul 19, 2021)

- Fix a bug which caused by missing gitalk.enable in config template.
- Now default using gitalk as comment.

## 🚀 v1.4.9.2 (Jul 16, 2021)

- Fix a bug which caused by missing "socials" in config template.
- Attention: you need to add at least one default social media list in config template.

## 🚀 v1.4.9.1 (May 20, 2021)

- Add aplayer to play remote musics. Now only support [NetEase Cloud Music](https://music.163.com). Add the code to theme
  config:

```yaml
# APlayer, now only support netease music list
aplayer:
    # Enable aplayer
    enable: true
    # Music list ID
    musiclist: ""
```

## 🚀 v1.4.9 (Apr 8, 2021)

- Add custom social platform for 3rdparty authors. This time support 3 platforms. You should add the following code in
  the fromt-meta of your article:

```yaml
socials:
    page:
        link: social platform link
        name: # the iconfont you want to use, for example: icon-bilibili-fill
        path: # the css file of your iconfont, for example: /css/font_bilibili/iconfont.css
    page2:
        link:
        name:
        path:
    page3:
        link:
        name:
        path: 
```

## 🚀 v1.4.8 (Apr 8, 2021)

- Add police beian
- Modify beian fomat
- Add the following code in your theme config file:

```yaml
# This function is only used for those that server is located in China. If your server is located in outside China, you do not need to add the following code.
#beian:
beian:
    # Enable
    enable: true
    # Beian location，for example：粤ICP备
    beianloc:
    # Beian ID，for example：2021022134号。
    beianid:

# Police Beian
police:
    # Enable
    enable: true
    # Beian location，for example: 粤公网安备
    beianloc:
    # Beian ID, for example: 44010602009049号
    beianid:
    # Beian Icon, for example: /img/beian.png
    beianicon:
```

## 🚀 v1.4.7 (Mar 31, 2021)

- Bug fix:
    - Now "preview" for the other article can correctly take effect now

## 🚀 v1.4.6 (Mar 31, 2021)

- Add beian ID：Add the following code in your theme config file:

```yaml
# This function is only used for those that server is located in China. If your server is located in outside China, you do not need to add the following code.
# Enable
beian: true
# Beian location，for example：粤ICP备
#beianloc:
# Beian ID，for example：2021022134号
#beianid:
```

- Custom 3rdparty author：add the following code in the front-matter of your markdown file：

```yaml
title: # example
author: # 3rdparty author name
is3rdparty: true
socials:
    page: # https://example
avatar: # /example.jpg
slogan: # example
```

- Custom gitalk proxy server to avoid 403 error，add the following code in your theme config file:

```yaml
...
# Upper is gitalk config code
proxy: https://cors-anywhere.azm.workers.dev/https://github.com/login/oauth/access_token
...
```

- Add custom mouse effect and custom mouse style, add the following code in your theme config file:

```yaml
mouse:
    # Enable custom mouse style
    enable: true
    # Default style
    default: # /default.png
    # Hover sytle
    pointer: # /hover.png
```

## 🚀 v1.4.5 (Jan 22, 2021)

### Improvements

- Changed the fonts links from using `loli.net` to using `baomitu.com` (which is hosted by 360, it's more stable). This
  changes will improve the website from loading up faster.

## 🚀 v1.4.4 (Sep 16, 2020)

Fixed couple of issues, get ready for some new features! Here thank you all for the 300+ stars so far, I will keep
working on improving this theme! ❤️

### Bug fixes 🐛

- Fixed when the `order list` text goes over one line, the number indicator and the text does not stay on the same line.
- Fixed when the `unorder list` text goes over one line, the dot indicator and the text does not stay on the same line.
- Fixed image caption repeat itself when the caption name repeats or it's empty.

## 🚀 v1.4.3 (Nov 13, 2019)

Added `MathJax` support! Also fixed a few issues here and there, smoothen the animation on article pages.

### Features

- Added `MathJax` support.

### Changes

- TOC now only loads up to level 3, this changes fixed the bug when TOC goes over 3 levels the content will overflowing
  out of the container. ([#20](https://github.com/TriDiamond/hexo-theme-obsidian/issues/20))
- Smoothen the animations for article pages.

## 🚀 v1.4.2 (Nov 11, 2019)

🎉 Obsidian theme now comes with search feature! Horray!
Hope you guys love this update!

### Features

- Added the mostly requested feature, local site search!

> The search feature is base on the index generated
> by [`hexo-generator-search`](https://github.com/wzpan/hexo-generator-search),
> So you need to install this plugin before you can use the search feature.

#### Installation guide

Install by npm

```bash
$ npm install hexo-generator-search --save
```

Adding the plugin config in your root `_config.yml`

```yaml
search:
    path: search.xml
    field: post
    content: true
```

Regenrate files

```bash
hexo clean && hexo g
```

Happy blogging! ❤️

## 🚀 v1.4.1 (Nov 9, 2019)

✨Shipped a complete redesigned `archives` page UI andenhanced modern `link style`!
Stay being awesome 👽!

### ✨Features

- Complete redesigned `Archives` page!
- Added fade in transition for all pages!
- Archives page now also allow custom header name config!

#### Check out the example theme config below

```yaml
# Page custom titles
page_titles:
    categories: 'Categories'
    tags: 'Tags'
    archives: 'Archived'
```

### 💥Changes

- New link styles for all pages!

source/live2d-widget/waifu-tips.js
/*
 * Live2D Widget
 * https://github.com/stevenjoezhang/live2d-widget
 */

function loadWidget(config) {
	let { waifuPath, apiPath, cdnPath } = config;
	let useCDN = false, modelList;
	//localStorage.setItem("Live2DPlace", "right");
	if (typeof cdnPath === "string") {
		useCDN = true;
		if (!cdnPath.endsWith("/")) cdnPath += "/";
	} else if (typeof apiPath === "string") {
		if (!apiPath.endsWith("/")) apiPath += "/";
	} else {
		console.error("Invalid initWidget argument!");
		return;
	}
	localStorage.removeItem("waifu-display");
	sessionStorage.removeItem("waifu-text");
	document.body.insertAdjacentHTML("beforeend", `<div id="waifu">
			<div id="waifu-tips"></div>
			<canvas id="live2d" width="800" height="800"></canvas>
			<div id="waifu-tool">
				<span class="fa fa-lg fa-comment"></span>
				<span class="fa fa-lg fa-paper-plane"></span>
				<span class="fa fa-lg fa-user-circle"></span>
				<span class="fa fa-lg fa-street-view"></span>
				<span class="fa fa-lg fa-camera-retro"></span>
				<span class="fa fa-lg fa-info-circle"></span>
				<span class="fa fa-lg fa-times"></span>
				<span class="fa fa-lg fa-chevron-right" id="live2d-go-right"></span>
                <span class="fa fa-lg fa-chevron-left" id="live2d-go-left"></span>
			</div>
		</div>`);
	// https://stackoverflow.com/questions/24148403/trigger-css-transition-on-appended-element
	setTimeout(() => {
		document.getElementById("waifu").style.bottom = 0;
	}, 0);

	function randomSelection(obj) {
		return Array.isArray(obj) ? obj[Math.floor(Math.random() * obj.length)] : obj;
	}
	// 检测用户活动状态，并在空闲时显示消息
	let userAction = false,
		userActionTimer,
		messageTimer,
		messageArray = ["好久不见，日子过得好快呢……", "大坏蛋！你都多久没理人家了呀，嘤嘤嘤～", "嗨～快来逗我玩吧！", "拿小拳拳锤你胸口！", "记得把小家加入 Adblock 白名单哦！"];
	window.addEventListener("mousemove", () => userAction = true);
	window.addEventListener("keydown", () => userAction = true);
	setInterval(() => {
		if (userAction) {
			userAction = false;
			clearInterval(userActionTimer);
			userActionTimer = null;
		} else if (!userActionTimer) {
			userActionTimer = setInterval(() => {
				showMessage(randomSelection(messageArray), 6000, 9);
			}, 20000);
		}
	}, 1000);

    (function () { // 根据位置加载
        if (localStorage.getItem("Live2DPlace") === "left") {
            document.getElementById("live2d-go-left").style.display = "none";
        } else if (localStorage.getItem("Live2DPlace") === "right") {
            document.getElementById("live2d-go-right").style.display = "none";
            document.getElementById("live2d_css").href = live2d_path + "waifu_right.css";
        }
    })();

	(function registerEventListener() {
		document.querySelector("#waifu-tool .fa-comment").addEventListener("click", showHitokoto);
		document.querySelector("#waifu-tool .fa-paper-plane").addEventListener("click", () => {
			if (window.Asteroids) {
				if (!window.ASTEROIDSPLAYERS) window.ASTEROIDSPLAYERS = [];
				window.ASTEROIDSPLAYERS.push(new Asteroids());
			} else {
				const script = document.createElement("script");
				script.src = "https://cdn.jsdelivr.net/gh/stevenjoezhang/asteroids/asteroids.js";
				document.head.appendChild(script);
			}
		});
		document.querySelector("#waifu-tool .fa-user-circle").addEventListener("click", loadOtherModel);
		document.querySelector("#waifu-tool .fa-street-view").addEventListener("click", loadRandModel);
		document.querySelector("#waifu-tool .fa-camera-retro").addEventListener("click", () => {
			showMessage("照好了嘛，是不是很可爱呢？", 6000, 9);
			Live2D.captureName = "photo.png";
			Live2D.captureFrame = true;
		});
		document.querySelector("#waifu-tool .fa-info-circle").addEventListener("click", () => {
			open("https://github.com/stevenjoezhang/live2d-widget");
		});
		document.querySelector("#waifu-tool .fa-times").addEventListener("click", () => {
			localStorage.setItem("waifu-display", Date.now());
			showMessage("愿你有一天能与重要的人重逢。", 2000, 11);
			document.getElementById("waifu").style.bottom = "-500px";
			setTimeout(() => {
				document.getElementById("waifu").style.display = "none";
				document.getElementById("waifu-toggle").classList.add("waifu-toggle-active");
			}, 3000);
		});
		
		document.querySelector("#waifu-tool .fa-chevron-right").addEventListener("click", () => { // 切换看板娘位置（左 => 右）
			localStorage.setItem("Live2DPlace", "right");
			showMessage("耶，可以去右边了呢～。", 2000, 11);
			document.getElementById("waifu").style.bottom = "-500px";
			document.getElementById("waifu-toggle").style.display = "none";
			setTimeout(() => {
				document.getElementById("live2d_css").href = live2d_path + "waifu_right.css";
				document.getElementById("waifu").style.bottom = "0px";
				document.getElementById("live2d-go-right").style.display = "none";
				document.getElementById("live2d-go-left").style.display = "block";
			}, 3000);
			setTimeout('document.getElementById("waifu-toggle").style.display = "inline"',6000);
		});
                document.querySelector("#waifu-tool .fa-chevron-left").addEventListener("click", () => { // 切换看板娘位置（左 <= 右）
			localStorage.setItem("Live2DPlace", "left");
			showMessage("耶，可以去左边了呢～。", 2000, 11);
			document.getElementById("waifu").style.bottom = "-500px";
			document.getElementById("waifu-toggle").style.display = "none";
			setTimeout(() => {
				document.getElementById("live2d_css").href = live2d_path + "waifu.css";
				document.getElementById("waifu").style.bottom = "0px";
				document.getElementById("live2d-go-left").style.display = "none";
				document.getElementById("live2d-go-right").style.display = "block";
			}, 3000);
			setTimeout('document.getElementById("waifu-toggle").style.display = "inline"',6000);
		});
		
		const devtools = () => {};
		console.log("%c", devtools);
		devtools.toString = () => {
			showMessage("哈哈，你打开了控制台，是想要看看我的小秘密吗？", 6000, 9);
		};
		window.addEventListener("copy", () => {
			showMessage("你都复制了些什么呀，转载要记得加上出处哦！", 6000, 9);
		});
		window.addEventListener("visibilitychange", () => {
			if (!document.hidden) showMessage("哇，你终于回来了～", 6000, 9);
		});
	})();

	(function welcomeMessage() {
		let text;
		if (location.pathname === "/") { // 如果是主页
			const now = new Date().getHours();
			if (now > 5 && now <= 7) text = "早上好！一日之计在于晨，美好的一天就要开始了。";
			else if (now > 7 && now <= 11) text = "上午好！工作顺利嘛，不要久坐，多起来走动走动哦！";
			else if (now > 11 && now <= 13) text = "中午了，工作了一个上午，现在是午餐时间！";
			else if (now > 13 && now <= 17) text = "午后很容易犯困呢，今天的运动目标完成了吗？";
			else if (now > 17 && now <= 19) text = "傍晚了！窗外夕阳的景色很美丽呢，最美不过夕阳红～";
			else if (now > 19 && now <= 21) text = "晚上好，今天过得怎么样？";
			else if (now > 21 && now <= 23) text = ["已经这么晚了呀，早点休息吧，晚安～", "深夜时要爱护眼睛呀！"];
			else text = "你是夜猫子呀？这么晚还不睡觉，明天起的来嘛？";
		} else if (document.referrer !== "") {
			const referrer = new URL(document.referrer),
				domain = referrer.hostname.split(".")[1];
			if (location.hostname === referrer.hostname) text = `欢迎阅读<span>「${document.title.split(" - ")[0]}」</span>`;
			else if (domain === "baidu") text = `Hello！来自 百度搜索 的朋友<br>你是搜索 <span>${referrer.search.split("&wd=")[1].split("&")[0]}</span> 找到的我吗？`;
			else if (domain === "so") text = `Hello！来自 360搜索 的朋友<br>你是搜索 <span>${referrer.search.split("&q=")[1].split("&")[0]}</span> 找到的我吗？`;
			else if (domain === "google") text = `Hello！来自 谷歌搜索 的朋友<br>欢迎阅读<span>「${document.title.split(" - ")[0]}」</span>`;
			else text = `Hello！来自 <span>${referrer.hostname}</span> 的朋友`;
		} else {
			text = `欢迎阅读<span>「${document.title.split(" - ")[0]}」</span>`;
		}
		showMessage(text, 7000, 8);
	})();

	function showHitokoto() {
		// 增加 hitokoto.cn 的 API
		fetch("https://v1.hitokoto.cn")
			.then(response => response.json())
			.then(result => {
				const text = `这句一言来自 <span>「${result.from}」</span>，是 <span>${result.creator}</span> 在 hitokoto.cn 投稿的。`;
				showMessage(result.hitokoto, 6000, 9);
				setTimeout(() => {
					showMessage(text, 4000, 9);
				}, 6000);
			});
	}

	function showMessage(text, timeout, priority) {
		if (!text || (sessionStorage.getItem("waifu-text") && sessionStorage.getItem("waifu-text") > priority)) return;
		if (messageTimer) {
			clearTimeout(messageTimer);
			messageTimer = null;
		}
		text = randomSelection(text);
		sessionStorage.setItem("waifu-text", priority);
		const tips = document.getElementById("waifu-tips");
		tips.innerHTML = text;
		tips.classList.add("waifu-tips-active");
		messageTimer = setTimeout(() => {
			sessionStorage.removeItem("waifu-text");
			tips.classList.remove("waifu-tips-active");
		}, timeout);
	}

	(function initModel() {
		let modelId = localStorage.getItem("modelId"),
			modelTexturesId = localStorage.getItem("modelTexturesId");
		if (modelId === null) {
			// 首次访问加载 指定模型 的 指定材质
			modelId = 1; // 模型 ID
			modelTexturesId = 53; // 材质 ID
		}
		loadModel(modelId, modelTexturesId);
		fetch(waifuPath)
			.then(response => response.json())
			.then(result => {
				window.addEventListener("mouseover", event => {
					for (let { selector, text } of result.mouseover) {
						if (!event.target.matches(selector)) continue;
						text = randomSelection(text);
						text = text.replace("{text}", event.target.innerText);
						showMessage(text, 4000, 8);
						return;
					}
				});
				window.addEventListener("click", event => {
					for (let { selector, text } of result.click) {
						if (!event.target.matches(selector)) continue;
						text = randomSelection(text);
						text = text.replace("{text}", event.target.innerText);
						showMessage(text, 4000, 8);
						return;
					}
				});
				result.seasons.forEach(({ date, text }) => {
					const now = new Date(),
						after = date.split("-")[0],
						before = date.split("-")[1] || after;
					if ((after.split("/")[0] <= now.getMonth() + 1 && now.getMonth() + 1 <= before.split("/")[0]) && (after.split("/")[1] <= now.getDate() && now.getDate() <= before.split("/")[1])) {
						text = randomSelection(text);
						text = text.replace("{year}", now.getFullYear());
						//showMessage(text, 7000, true);
						messageArray.push(text);
					}
				});
			});
	})();

	async function loadModelList() {
		const response = await fetch(`${cdnPath}model_list.json`);
		modelList = await response.json();
	}

	async function loadModel(modelId, modelTexturesId, message) {
		localStorage.setItem("modelId", modelId);
		localStorage.setItem("modelTexturesId", modelTexturesId);
		showMessage(message, 4000, 10);
		if (useCDN) {
			if (!modelList) await loadModelList();
			const target = randomSelection(modelList.models[modelId]);
			loadlive2d("live2d", `${cdnPath}model/${target}/index.json`);
		} else {
			loadlive2d("live2d", `${apiPath}get/?id=${modelId}-${modelTexturesId}`);
			console.log(`Live2D 模型 ${modelId}-${modelTexturesId} 加载完成`);
		}
	}

	async function loadRandModel() {
		const modelId = localStorage.getItem("modelId"),
			modelTexturesId = localStorage.getItem("modelTexturesId");
		if (useCDN) {
			if (!modelList) await loadModelList();
			const target = randomSelection(modelList.models[modelId]);
			loadlive2d("live2d", `${cdnPath}model/${target}/index.json`);
			showMessage("我的新衣服好看嘛？", 4000, 10);
		} else {
			// 可选 "rand"(随机), "switch"(顺序)
			fetch(`${apiPath}rand_textures/?id=${modelId}-${modelTexturesId}`)
				.then(response => response.json())
				.then(result => {
					if (result.textures.id === 1 && (modelTexturesId === 1 || modelTexturesId === 0)) showMessage("我还没有其他衣服呢！", 4000, 10);
					else loadModel(modelId, result.textures.id, "我的新衣服好看嘛？");
				});
		}
	}

	async function loadOtherModel() {
		let modelId = localStorage.getItem("modelId");
		if (useCDN) {
			if (!modelList) await loadModelList();
			const index = (++modelId >= modelList.models.length) ? 0 : modelId;
			loadModel(index, 0, modelList.messages[index]);
		} else {
			fetch(`${apiPath}switch/?id=${modelId}`)
				.then(response => response.json())
				.then(result => {
					loadModel(result.model.id, 0, result.model.message);
				});
		}
	}
}

function initWidget(config, apiPath) {
	if (typeof config === "string") {
		config = {
			waifuPath: config,
			apiPath
		};
	}
	document.body.insertAdjacentHTML("beforeend", `<div id="waifu-toggle">
			<span>看板娘</span>
		</div>`);
	const toggle = document.getElementById("waifu-toggle");
	toggle.addEventListener("click", () => {
		toggle.classList.remove("waifu-toggle-active");
		if (toggle.getAttribute("first-time")) {
			loadWidget(config);
			toggle.removeAttribute("first-time");
		} else {
			localStorage.removeItem("waifu-display");
			document.getElementById("waifu").style.display = "";
			setTimeout(() => {
				document.getElementById("waifu").style.bottom = 0;
			}, 0);
		}
	});
	if (localStorage.getItem("waifu-display") && Date.now() - localStorage.getItem("waifu-display") <= 86400000) {
		toggle.setAttribute("first-time", true);
		setTimeout(() => {
			toggle.classList.add("waifu-toggle-active");
		}, 0);
	} else {
		loadWidget(config);
	}
}
